src/plugins/discover/public/application/main/components/layout/discover_documents.tsx

```diff
@@ -71,34 +74,104 @@ export const onResize = (
+        rows: documents$.getValue().result,
+      })
+  );
+  useEffect(() => {
+    const subscribe = merge(stateContainer.savedSearchState.getCurrent$(), documents$).subscribe(
+      (next) => {
+        const currentState = state$.getValue();
+        if (!isSavedSearch(next)) {
+          const dataViewLoading = currentState.dataView !== next.dataView;
+          if (
+            dataViewLoading ||
+            (next.fetchStatus === FetchStatus.LOADING &&
+              !currentState.isDataLoading &&
+              !next.result)
+          ) {
+            state$.next({ ...state$.getValue(), isDataLoading: true });
+          }
+
+          if (next.fetchStatus === FetchStatus.COMPLETE) {
+            const nextMsg = {
+              ...state$.getValue(),
+              isDataLoading: false,
+              rows: next.result,
+            };
+            if (next.fetchAppState) {
+              nextMsg.appState = { ...currentState.appState, ...next.fetchAppState };
+            }
+            if (next.dataView) {
+              nextMsg.dataView = next.dataView;
+            }
+            state$.next(nextMsg);
+          }
+        } else {
+          if (currentState.dataView.id !== next.searchSource.getField('index')?.id) {
+            return;
+          }
+          // update appState by savedSearch unless the data view change is ongoing ... too much updates else
+
+          const nextAppState = updateAppStateBySavedSearch({
+            savedSearch: next,
+            appState: currentState.appState,
+            uiSettings: discoverServices.uiSettings,
+            updateDataView: false,
+          });
+
+          if (!isEqual(nextAppState, currentState?.appState)) {
+            state$.next({ ...state$.getValue(), appState: nextAppState });
+          }
+        }
+      }
+    );
+    return () => {
+      subscribe.unsubscribe();
+    };
+  }, [stateContainer, state$, documents$, discoverServices.uiSettings]);
+  return useObservable(state$, state$.getValue());
+}
+
 function DiscoverDocumentsComponent({
-  dataView,
   onAddFilter,
   stateContainer,
   onFieldEdited,
 }: {
-  dataView: DataView;
-  onAddFilter?: DocViewFilterFn;
+  onAddFilter: DocViewFilterFn;
   stateContainer: DiscoverStateContainer;
-  onFieldEdited?: () => void;
+  onFieldEdited: () => void;
 }) {
   const services = useDiscoverServices();
-  const documents$ = stateContainer.dataState.data$.documents$;
-  const savedSearch = useSavedSearchInitial();
   const { dataViews, capabilities, uiSettings } = services;
-  const [query, sort, rowHeight, rowsPerPage, grid, columns, index] = useAppStateSelector(
-    (state) => {
-      return [
-        state.query,
-        state.sort,
-        state.rowHeight,
-        state.rowsPerPage,
-        state.grid,
-        state.columns,
-        state.index,
-      ];
-    }
+  const { dataView, appState, isDataLoading, rows } = useSavedSearchDataState(
+    stateContainer,
+    services
```

Comment:
the second significant change, use a dedicated saved search data cache to reduce updates and re-renderings of this component